### ISSUE TITLE
mypy doesn't detect None checks

### ISSUE THREAD

**Issue by VoidSignal** [python/mypy#7268]
Hi, mypy docs say that it can detect `None` checks, but that doesn't work with my example below.

>* Are you reporting a bug, or opening a feature request?

I believe this is a bug

>* Please insert below the code you are checking with mypy,
```Python
from typing import Callable, Optional, Union

Number = Union[int, float]
Func = Callable[[Number], Number]


def composition(double: Func,
                square: Optional[Func] = None,
                inc: Optional[Func] = None) -> Func:
    if square:
        if inc:
            return lambda x: double(square(inc(x)))
        else:
            return lambda x: double(square(x))
    else:
        return double
```
>* What is the actual behavior/output?
```
temp.py:12: error: "None" not callable
temp.py:14: error: "None" not callable
```
>* What is the behavior/output you expect?

Should type check without an issue.
>* What are the versions of mypy and Python you are using?

0.711

>Do you see the same issue after installing mypy from Git master?

Yes

* What are the mypy flags you are using? (For example --strict-optional)

`--strict`

**Comment by tavianator** [comment]
I believe the reason this happens is that the lambda captures the variable by reference, and it could hypothetically be modified after the lambda is created, so it must be conservative about the type.  You can write something like
```
if square:
    _square = square
    if inc:
        _inc = inc
        return lambda x: double(_square(_inc(x)))
...
```

**Comment by ilevkivskyi** [comment]
Yes, this is a duplicate of https://github.com/python/mypy/issues/2608
